torch/_inductor/fx_passes/quantization.py

```diff
@@ -142,3 +164,196 @@ def qconv(match: Match, *args, **kwargs):
+        pattern,
+        extra_check=_is_valid_dequant_conv2d_pattern,
+        pass_number=1,  # pass_number=1, ensure it's behand dequant promotion pass
+    )
+    def qconv_weight_prepack(match: Match, *args, **kwargs):
+        """
+        Macth the pattern:
+        int8 activation
+          |
+        dequant_per_tensor
+          |
+        Conv2d <- optional(aten.clone.default) <- dequant_per_channel
+
+        Insert weight prepack node and change the pattern to:
+        int8 activation
+          |
+        onednn.qconv2d_pointwise <- onednn.qconv_prepack <- optional(aten.clone.default) <- dequant_per_channel
+        """
+        has_clone_to_channel_last_node_in_pattern = any(
+            node.target == aten.clone.default for node in match.nodes
+        )
+
+        (
+            to_fp32_node,
+            sub_node,
+            mul_node,
+            dequant_per_channel,
+        ) = match.nodes[0:4]
+
+        clone_node = (
+            match.nodes[4] if has_clone_to_channel_last_node_in_pattern else None
+        )
+        conv_node = (
+            match.nodes[5]
+            if has_clone_to_channel_last_node_in_pattern
+            else match.nodes[4]
+        )
+        if clone_node is not None:
+            assert clone_node.target is aten.clone.default
+        assert conv_node.target is aten.convolution.default
+
+        # Activation QParams
+        qx, x_zp, x_scale = (
+            kwargs["x"],
+            kwargs["x_zp"],
+            kwargs["x_scale"],
+        )
+
+        # Weight QParams
+        qw, w_scale, w_zp = (
+            kwargs["q_weight"],
+            kwargs["w_scale"],
+            kwargs["w_zp"],
+        )
+
+        # Conv Params
+        bias, stride, padding, dilation, groups = (
+            kwargs["b"],
+            kwargs["stride"],
+            kwargs["padding"],
+            kwargs["dilation"],
+            kwargs["groups"],
+        )
+
+        x_shape = qx.meta.get("tensor_meta").shape
+        graph = match.graph
+        with graph.inserting_before(conv_node):
+            # Insert weight prepack node and the QConv node
+            packed_weight_inputs = (
+                qw,
+                w_scale,
+                x_scale,
+                x_zp,
+                stride,
+                padding,
+                dilation,
+                groups,
+                x_shape,
+            )
+            packed_weight_op = torch.ops.onednn.qconv_prepack
+            prepack_weight_node = graph.call_function(
+                packed_weight_op, args=packed_weight_inputs
+            )
+
+            new_args = (
+                qx,
+                x_scale,
+                x_zp,
+                prepack_weight_node,
+                w_scale,
+                w_zp,
+                bias,
+                stride,
+                padding,
+                dilation,
+                groups,
+                1.0,  # output_scale
+                0,  # output_zero_point
```

Comment:
This is 1 and 0 because the op being inserted generated fp32 output that gets quantized separately? and later you do appropriate fusion?

Reply:
Yean, here we fuse it into the conv which generated fp32 output. Then, we will further check if there is another quant pattern at output in the following passes.